File: KP_Methods/C++/Element_Based.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

int knapsack_by_elements(vector<int>& elements, vector<vector<int>>& items, vector<int>& profits, int max_weight) {
    int n = elements.size();
    int m = items.size();

    // Calculate the weight of each item and store it in a map
    unordered_map<int, int> item_weights;
    for (int i = 0; i < m; i++) {
        int weight = 0;
        for (int j = 0; j < n; j++) {
            if (items[i][j] == 1) {
                weight += elements[j];
            }
        }
        item_weights[i] = weight;
    }

    // Sort the items in decreasing order of weight and store their indices in a vector
    vector<int> item_indices(m);
    for (int i = 0; i < m; i++) {
        item_indices[i] = i;
    }
    sort(item_indices.begin(), item_indices.end(), [&](int i, int j) {
        return item_weights[i] > item_weights[j];
    });

    // Initialize a map to keep track of the remaining elements for each item
    unordered_map<int, vector<int>> remaining_elements;
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (items[i][j] == 1) {
                remaining_elements[i].push_back(j);
            }
        }
    }

    int total_profit = 0;

    // Iterate through the items in decreasing order of weight
    for (int i : item_indices) {
        // If the weight of the current item is greater than the remaining capacity,
        // skip the item
        if (item_weights[i] > max_weight) {
            continue;
        }

        // Iterate through the remaining elements of the current item
        for (int j = 0; j < remaining_elements[i].size(); j++) {
            int element = remaining_elements[i][j];

            // If the weight of the current element is greater than the remaining capacity,
            // skip the element
            if (elements[element] > max_weight) {
                continue;
            }

            // Add the profit of the current item to the total profit
            total_profit += profits[i];

            // Update the remaining capacity
            max_weight -= elements[element];

            // Remove the current element from the remaining elements of all items
            for (auto& kv : remaining_elements) {
                auto& elements = kv.second;
                elements.erase(remove(elements.begin(), elements.end(), element), elements.end());
            }

            // Remove the current item from the list of remaining items
            item_indices.erase(remove(item_indices.begin(), item_indices.end(), i), item_indices.end());

            // Decrement the index and break out of the loop
            j--;
            break;
        }
    }

    return total_profit;
}

int main() {
    vector<int> elements = {10, 20, 30};
    vector<vector<int>> items = {{1, 1, 0}, {1, 0, 1}, {0, 1, 1}};
    vector<int> profits = {60, 100, 120};
    int max_weight = 50;

    cout << knapsack_by_elements(elements, items, profits, max_weight) << endl;

    return 0;
}

File: KP_Methods/Python/Element_Based.py
def knapsack_by_elements(elements, items, profits, max_weight):
    # Calculate the weights of the elements and the number of elements in each item
    element_weights = [0] * len(elements)
    element_count = [0] * len(elements)
    for i, item in enumerate(items):
        for j, elem in enumerate(item):
            if elem == 1:
                element_weights[j] += elements[j]
                element_count[j] += 1

    # Sort the elements by their weight in descending order
    sorted_elements = sorted(range(len(elements)), key=lambda k: -element_weights[k])

    # Initialize the knapsack and the total profit
    knapsack = [0] * len(items)
    total_profit = 0

    # Iterate through the sorted elements and add them to the knapsack
    for elem in sorted_elements:
        # Skip elements that are already in the knapsack or that are too heavy
        if element_count[elem] == 0 or elements[elem] > max_weight:
            continue

        # Find the items that contain this element and sort them by their remaining weight
        eligible_items = [(i, sum([elements[k] for k, e in enumerate(items[i]) if e == 1])) for i in range(len(items)) if items[i][elem] == 1]
        eligible_items.sort(key=lambda x: x[1])

        # Iterate through the eligible items and add them to the knapsack
        for item in eligible_items:
            item_idx = item[0]
            item_weight = item[1]
            if item_weight <= max_weight:
                # Add the item to the knapsack
                knapsack[item_idx] = 1
                element_count[elem] -= 1
                max_weight -= item_weight
                total_profit += profits[item_idx]
                break

    return total_profit

# Example usage
elements = [10, 20, 30]
items = [[1, 1, 0], [1, 0, 1], [0, 1, 1]]
profits = [60, 100, 120]
max_weight = 50
print(knapsack_by_elements(elements, items, profits, max_weight))

import pandas as pd
pd.options.mode.chained_assignment = None  # default='warn'
import numpy as np

# Selection of instance
data = pd.read_csv(r'C:\Users\Javier\Documents\IA\Fondecyt\KP_Methods\Instances\InstancesofSetI\85_100_0.10_0.75.txt', header=None)

n_items = data[0][0]
n_elements = data[0][1]
max_weight = int(data[0][2])

profits = str(data[0][3]).split()
for i in range(len(profits)):
    profits[i] = int(profits[i])

elements = str(data[0][4]).split()
for i in range(len(elements)):
    elements[i] = int(elements[i])

items = data.tail(-5)
items.reset_index(inplace = True, drop = True)

for i in range(len(items)):
    items[0][i] = str(items[0][i]).split()
for i in range(len(items)):
    for j in range(len(elements)):
        items[0][i][j] = int(items[0][i][j])

items = items[0].values.tolist()

print(knapsack_by_elements(elements, items, profits, max_weight))
